File: src/ruby/tormenta20/models/melhoria.rb
# frozen_string_literal: true

require "active_record"
require "json"

module Tormenta20
  module Models
    # Model representing a Melhoria (Enhancement/Upgrade) in Tormenta20.
    #
    # Enhancements can be applied to weapons, armor, and other items
    # to provide magical bonuses and special effects.
    #
    # @example Find all enhancements
    #   Tormenta20.melhorias.all
    #
    # @example Find a specific enhancement
    #   Tormenta20.melhorias.find("flamejante")
    #
    # @!attribute [rw] id
    #   @return [String] Unique identifier (e.g., "flamejante")
    # @!attribute [rw] name
    #   @return [String] Enhancement name in Portuguese
    # @!attribute [rw] description
    #   @return [String, nil] Enhancement description
    # @!attribute [rw] applicable_to
    #   @return [Array<String>, nil] Item types this enhancement can be applied to
    # @!attribute [rw] price
    #   @return [String, nil] Price or price modifier for this enhancement
    # @!attribute [rw] effects
    #   @return [Array<String>, nil] Effects provided by this enhancement
    class Melhoria < Base
      self.table_name = "melhorias"

      validates :id, presence: true, uniqueness: true
      validates :name, presence: true

      # Convert the enhancement to a hash representation.
      #
      # @return [Hash] Hash with all enhancement attributes (nil values excluded)
      def to_h
        {
          id: id,
          name: name,
          description: description,
          applicable_to: applicable_to,
          price: price,
          effects: effects
        }.compact
      end
    end
  end
end

File: src/ts/models/melhoria.ts
import { BaseModel } from "./base.js"
import { Query } from "../query.js"

const TABLE = "melhorias"

/**
 * A Melhoria (item upgrade/improvement) in Tormenta20.
 *
 * Melhorias are purchased enhancements applied to existing equipment.
 *
 * @example
 * ```ts
 * Melhoria.all()
 * Melhoria.find("resistente")
 * ```
 */
export class Melhoria extends BaseModel {
  get description(): string | null { return (this._row.description as string) ?? null }
  /** JSON array of equipment types this melhoria can be applied to. */
  get applicableTo(): unknown[] { return BaseModel.json(this._row.applicable_to, []) }
  /** Price in tibares. */
  get price(): number { return (this._row.price as number) ?? 0 }
  /** JSON object with mechanical effects. */
  get effects(): Record<string, unknown> { return BaseModel.json(this._row.effects, {}) }

  toH() {
    return {
      id: this.id, name: this.name, description: this.description,
      applicable_to: this.applicableTo, price: this.price, effects: this.effects,
    }
  }

  static query(): Query<Melhoria> { return BaseModel.makeQuery(TABLE, Melhoria) }
  static all(): Melhoria[] { return Melhoria.query().all() }
  static find(id: string): Melhoria | null { return Melhoria.query().find(id) }
  static first(): Melhoria | null { return Melhoria.query().first() }
  static last(): Melhoria | null { return Melhoria.query().last() }
  static count(): number { return Melhoria.query().count() }
}
